pkg/quarterdeck/projects.go

```diff
@@ -0,0 +1,271 @@
+package quarterdeck
+
+import (
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/oklog/ulid/v2"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/api/v1"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/db/models"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/middleware"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/permissions"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/responses"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/tokens"
+	"github.com/rotationalio/ensign/pkg/utils/metrics"
+	"github.com/rotationalio/ensign/pkg/utils/pagination"
+	"github.com/rotationalio/ensign/pkg/utils/sentry"
+	"github.com/rotationalio/ensign/pkg/utils/ulids"
+)
+
+func (s *Server) ProjectList(c *gin.Context) {
+	var (
+		err      error
+		orgID    ulid.ULID
+		out      *api.ProjectList
+		claims   *tokens.Claims
+		projects []*models.Project
+		prevPage *pagination.Cursor
+		nextPage *pagination.Cursor
+	)
+
+	// Fetch the user claims from the reuqest
+	if claims, err = middleware.GetClaims(c); err != nil {
+		sentry.Error(c).Err(err).Msg("could not get user claims from authenticated request")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse(responses.ErrSomethingWentWrong))
+		return
+	}
+
+	if orgID = claims.ParseOrgID(); ulids.IsZero(orgID) {
+		sentry.Error(c).Msg("could not parse orgID from claims")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse(responses.ErrSomethingWentWrong))
+		return
+	}
+
+	// Process the next page token or page query options
+	query := &api.PageQuery{}
+	if err = c.BindQuery(query); err != nil {
+		sentry.Warn(c).Err(err).Msg("could not parse api page query")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse(api.ErrUnparsable))
+		return
+	}
+
+	if query.NextPageToken != "" {
+		if prevPage, err = pagination.Parse(query.NextPageToken); err != nil {
+			sentry.Warn(c).Err(err).Str("next_page_token", query.NextPageToken).Msg("could not parse next page token")
+			c.JSON(http.StatusBadRequest, api.ErrorResponse(err))
+			return
+		}
+	} else {
+		prevPage = pagination.New("", "", int32(query.PageSize))
+	}
+
+	if projects, nextPage, err = models.ListProjects(c.Request.Context(), orgID, prevPage); err != nil {
+		sentry.Error(c).Err(err).Msg("could not parse orgID from claims")
```

Comment:
Should this have a different error message?

Reply:
good catch!